testing/integration-testing-hooks.ts

```diff
@@ -1,92 +1,398 @@
+    for (const ext of ['', '.exe', '.bat']) {
+      try {
+        await fs.stat(mlaunchPy + ext);
+        return mlaunchPath = {
+          exec: mlaunchPy + ext,
+          env: { PYTHONPATH: tmpdir }
+        };
+      } catch {}
+    }
+  } catch {}
+
+  await execFile('pip3', ['install', '--user', 'mtools[mlaunch]']);
+  // Figure out the most likely target path for pip3 --user and use mlaunch
+  // from there.
+  const pythonBase = (await execFile(python, ['-m', 'site', '--user-base'])).stdout.trim();
+  const pythonPath = (await execFile(python, ['-m', 'site', '--user-site'])).stdout.trim();
+  const mlaunchExec = path.join(pythonBase, 'bin', 'mlaunch');
+  let lastErr = new Error('unreachable');
+  for (const ext of ['', '.exe', '.bat']) {
+    try {
+      await fs.stat(mlaunchExec + ext);
+      return mlaunchPath = {
+        exec: mlaunchExec + ext,
+        env: { PYTHONPATH: pythonPath }
+      };
+    } catch(err) {
+      lastErr = err;
+    }
+  }
+  throw lastErr;
+}
+
+type MlaunchCommand = 'init' | 'start' | 'stop' | 'list' | 'kill';
+// Run a specific mlaunch command, trying to ensure that `mlaunch` is accessible
+// and installing it if necessary.
+async function execMlaunch(command: MlaunchCommand, ...args: string[]): Promise<void> {
+  const mlaunchPath = await getMlaunchPath();
+
+  // command could be part of args, but the extra typechecking here has helped
+  // me at least once.
+  const fullCmd = [mlaunchPath.exec, command, ...args];
+  console.info('Running command', fullCmd.join(' '));
+  const proc = child_process.spawn(fullCmd[0], fullCmd.slice(1), {
+    env: { ...process.env, ...mlaunchPath.env },
+    stdio: 'inherit'
+  });
+  await once(proc, 'exit');
+  console.info('Successfully ran command', fullCmd.join(' '));
+}
+
+// Remove all potential leftover mlaunch instances.
+export async function clearMlaunch() {
+  const tmpdir = await getTmpdir();
+  for await (const { name } of await fs.opendir(tmpdir)) {
+    if (name.startsWith('mlaunch-')) {
+      const fullPath = path.join(tmpdir, name);
+      try {
+        await execMlaunch('kill', '--dir', fullPath);
+        await promisify(rimraf)(fullPath);
+      } catch (err) {
+        console.warn(`Removing ${fullPath} failed:`, err);
+      }
+    }
+  }
+  await promisify(rimraf)(path.join(tmpdir, '.current-port'));
+}
+
+// Represents one running test server instance.
+export class MongodSetup {
+  _connectionString: string;
+  _serverVersion: string | null = null;
+
+  constructor(connectionString: string) {
+    this._connectionString = connectionString;
+  }
+
+  async start(): Promise<void> {
+    throw new Error('Server not managed');
+  }
+
+  async stop(): Promise<void> {
+    throw new Error('Server not managed');
+  }
+
+  connectionString(): string {
+    return this._connectionString;
+  }
+
+  host(): string {
+    return new URL(this.connectionString()).hostname;
+  }
+
+  port(): string {
```

Comment:
I'm not sure I understand -- the default port for `mongodb://` URLs is `27017`, so we should return that if there's no port in the URL, right?

Reply:
Yes - that's right. I was thinking maybe we'd want to prevent tests from starting a server on 27017 (since most people have something else running on 27017 so if you start your test processes it may cause collisions), but it looks like none of the tests are doing that anyway and it would be incorrect/unexpected to always avoid 27017. So ignore my comment!

Reply:
The general idea here is that you can still specify `MONGOSH_TEST_SERVER_URL` for debugging if that helps to point to a specific server, and that in that case, if that URL does not have a port, we should use `27017` :)